### PR TITLE
Add sops lookup plugin

### LINKED ISSUE #47770: Integration with Mozilla SOPS for encrypted files
##### SUMMARY

Integrate Mozilla `sops` to allow file content lookup.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

Lookup plugin

##### ADDITIONAL INFORMATION

Issue #36982 requests `sops` integration for encrypted vars file.

As `sops` is capable of encrypting/decrypting [entire files][entire-files], a `lookup` plugin to load them can come useful to use encrypted file content.

[entire-files]: https://github.com/mozilla/sops#operational-requirements

As an example:

```yaml
tasks:
- name: add ssh private key
  copy:
    content: "{{lookup('sops', user + '-id_rsa')}}"
    dest: /home/{{user}}/.ssh/id_rsa
    owner: "{{user}}"
    group: "{{user}}"
    mode: 0600
```

### PR DESCRIPTION
**Pull request by endorama**
##### SUMMARY

Add a new lookup plugin for using [`sops`](https://github.com/mozilla/sops) to lookup content from sops-encrypted files.

Fixes #47770

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

plugin 

##### ADDITIONAL INFORMATION

Sops can be used in place of Ansible Vault as an easy to manage at rest encryption for secrets in the repository.

Sops is able to encrypt entire files, this plugin allow making advantage of that functionality and load as strings sops encrypted files (think of TLS cert keys, SSH keys or any other entire file secret you may need).

This plugin require the `sops` command to be available in the path.

Related to #36982

### CODE REVIEW

File: lib/ansible/plugins/lookup/sops.py
Comment (jwitko) [inline lib/ansible/plugins/lookup/sops.py]: This variable is causing an `undefined variable` error.
```suggestion
                    command = ["sops", "--decrypt", lookupfile]
```

**Review by jwitko — CHANGES_REQUESTED** [review]
I went looking hoping that I'd find something just like this.  Thank you very much for contributing!  I have tested this and was able to make it work quite nice with the suggested edit below, although I did have to `| from_json` in order to get usable output.  Not sure if its desirable to handle that in the module or let the user manipulate the output?

Example:
Play:
```
---
- hosts: local
  gather_facts: False
  tasks:
    - copy:
        content: "{{ lookup('sops', 'enc.json') }}"
        dest: /tmp/tmp
```

output:
```
$ cat /tmp/tmp
b'{\n\t"auths": {\n\t\t"iad.ocir.io": {\n\t\t\t"username": "TEST/TEST",\n\t\t\t"password": "nn\\u003c+0xX55V;V\\u003eRjusAew",\n\t\t\t"email": "redacted@email.com",\n\t\t\t"auth": "TEST"\n\t\t}\n\t}\n}'
```

output when adding `| from_json` to the `content:` line:
```
{"auths": {"iad.ocir.io": {"username": "TEST/TEST", "password": "nn<+0xX55V;V>RjusAew", "email": "redacted@email.com", "auth": "TEST"}}}
```

You can see how the output is completely unusable without `| from_json` so I'm kind of having a hard time thinking of a reason why we wouldn't want it parsed to a usable format from inside the module?

File: lib/ansible/plugins/lookup/sops.py
Comment (jwitko) [reply lib/ansible/plugins/lookup/sops.py]: @endorama ping

Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: I pushed the fix!

**Comment by endorama** [comment]
@jwitko I fixed pylint and the undefined var. (I'll squash commits after review if you are ok)

Regaring the json not being parsed, this is probably a bug.

Let me clarify this a bit. Sops can operate in 2 "modes": structured or unstructured file.  
Structured files are json and yaml, you expect that to be parsed before use.  
Unstructured files are any binary or textual file that you expect to use as is after decryption.

I imagined the lookup plugin to be useful for unstructured files mainly. For example certificates private keys. In this case you use the output as it is, and you expect no transformation to happen.

Instead, you could use the vars plugin, proposed in PR #59641, to load playbook variables from structured files variables.

Your particular use case seems to be broken by some formatting issue in the string read from the sops output. I replicated the behaviour and trying to fix it, as it should not happen.

**Comment by endorama** [comment]
Commit 0a5924d fixes the issue showed by @jwitko, which I guess was due to changes in how `process.communicate()` handles command output (maybe Python 3 related?)

I also added 2 plugin variables:
- `binary` to control if the `sops` command output should handled as binary or text (default text)
- `print` to control if the decrypted secret should be printed on screen or not (instead of commenting/decommenting the code)

File: lib/ansible/plugins/lookup/sops.py
Comment (sivel) [inline lib/ansible/plugins/lookup/sops.py]: Lookups should always return "text".  As such, you should make use of `ansible.module_utils._text.to_text` to handle enforce the type.
```
output = to_text(output, errors='surrogate_or_strict')
```

Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: thanks for the suggestion! I changed the code to use the proper ansible function.

**Comment by endorama** [comment]
ready_for_review

**Review by jwitko — APPROVED** [review]

**Comment by jwitko** [comment]
@sivel Not sure how to make the CI retry but it looks like it failed due to a transient issue in the tests not an actual issue in the code.

**Comment by endorama** [comment]
Shall I squash this in a single commit?

**Comment by cameron-elliott** [comment]
Hey! I just checked out this PR, which means it must be pretty cool, because I struggled a fair bit being new to gpg, sops, ansible and Python. :smile:
If I can provide any feedback or newbie-level help, let me know, but I think having this available is pretty neato, thanks for contributing it!

**Comment by jwitko** [comment]
@sivel seems we're a but stuck here

**Comment by mhumeSF** [comment]
Am I running this correctly? Using the code from the example in `sops.py`

```
---
  - name: output secrets to screen (BAD IDEA)
    debug:
        msg: "Content: {{lookup('sops', item)}}"
    with_file:
    - secrets.enc.yaml
```
Getting errors when running:
```
Sops lookup using None as file
fatal: [localhost]: FAILED! => {
    "msg": "An unhandled exception occurred while running the lookup plugin 'sops'. Error was a <class 'ansible.errors.AnsibleError'>, original message: could not locate file in lookup: password: ENC[AES256_GCM,data:redacted,tag:redacted,type:str] sops: kms: -   arn: arn:aws:kms:<redacted> created_at: '2019-10-10T21:18:01Z' enc: redacted aws_profile: "" gcp_kms: [] azure_kv: [] lastmodified: '2019-10-10T21:18:02Z' mac: ENC[AES256_GCM,data:redacted,iv:redacted,tag:redacted,type:str] pgp: [] unencrypted_suffix: _unencrypted version: 3.4.0
```

**Comment by endorama** [comment]
@mhumeSF thanks for spotting this! That's an error in the documentation.

It uses `with_file`, but as it returns the file *content* `sops` breaks, thus the error you see. I fixed the documentation, but using `with_items` (so passing the *path* to the lookup) should make it work.

File: lib/ansible/plugins/lookup/sops.py
Comment (mhumeSF) [inline lib/ansible/plugins/lookup/sops.py]: Would this be better? `process = Popen(command, stdout=PIPE, stderr=PIPE, encoding='utf8')`

**Comment by mhumeSF** [comment]
@endorama Updated with `with_items`, which helped. Now the file was found but, was getting a new error of `Error was a <class 'TypeError'>, original message: a bytes-like object is required, not 'str'"`.

File: lib/ansible/plugins/lookup/sops.py
Comment (mhumeSF) [reply lib/ansible/plugins/lookup/sops.py]: This was due to an environment issue on my end, so disregard.

**Comment by mhumeSF** [comment]
@sivel These two plugins are working great. What else would be needed to get this and https://github.com/ansible/ansible/pull/59641 merged?

**Review by mhumeSF — APPROVED** [review]
👍

**Comment by endorama** [comment]
I'm failing to see where it breaks, any hint appreciated! (Maybe rebase onto devel would help?)

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]:
```suggestion
    version_added: "2.10"
```

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]:
```suggestion
        - This lookup requires the C(sops) executable to be available in the controller PATH.
```

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]:
```suggestion
        - This lookup does not understand 'globing' - use the fileglob lookup instead.
```

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]: You want to move that to module_utils once at least one of this and #59641 is merged.

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]: What is this option needed for? After all, these are secrets you're dealing with, and if the user wants to print them she can use the `debug` module.

File: lib/ansible/plugins/lookup/sops.py
Comment (mweinelt) [reply lib/ansible/plugins/lookup/sops.py]: Globing -> Globbing?

Comment (felixfontein) [reply lib/ansible/plugins/lookup/sops.py]: Indeed :) I totally missed that...

File: lib/ansible/plugins/lookup/sops.py
Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: You're perfectly right, actually I didn't thought about that. I'll remove it.

File: lib/ansible/plugins/lookup/sops.py
Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]:
```suggestion
        - This lookup does not understand 'globbing' - use the fileglob lookup instead.
```

File: lib/ansible/plugins/lookup/sops.py
Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: I will surely need some guidance on that, as I already tried but couldn't make it work. But sure, happy to tackle that.

File: lib/ansible/plugins/lookup/sops.py
Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: As suggested, see a262523

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [inline lib/ansible/plugins/lookup/sops.py]:
```suggestion
    author: Edoardo Tenani (@endorama) <e.tenani@arduino.cc>
```
(Same for the other PR. This will make sure that you will be notified about issues / PRs for this plugin by ansibot.)

File: lib/ansible/plugins/lookup/sops.py
Comment (felixfontein) [reply lib/ansible/plugins/lookup/sops.py]: Essentially, it works as follows:
 1. Create a file `lib/ansible/module_utils/sops.py` which starts with the following, and add the code which is used by both plugins (i.e. `sops_error_codes`, `SopsError`, `Sops` from this PR):
```
# Copyright (c), xxx <yyy@zzz.com>, 2018-2020
# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)

from __future__ import absolute_import, division, print_function
```
2. Add `from ansible.module_utils.sops import Sops, SopsError` to the plugin (and remove the code you've copied).

**Review by felixfontein — COMMENTED** [review]
Besides this and creating a module_utils, you should also add tests for this. As an example, you can look at:
- https://github.com/ansible/ansible/tree/devel/test/integration/targets/lookup_rabbitmq
- https://github.com/ansible/ansible/tree/devel/test/integration/targets/setup_rabbitmq

(You will be able to re-use `setup_sops` for #59641.)

File: lib/ansible/plugins/lookup/sops.py
Comment (endorama) [reply lib/ansible/plugins/lookup/sops.py]: Would you prefer this change to be in another PR or it's fine to add it to this one?

Comment (felixfontein) [reply lib/ansible/plugins/lookup/sops.py]: I think it will be faster to merge this PR (since the other has the paths issue), so let's make the change here and get this merged, and adjust the other PR according (once this is merged). Does that sound good for you?

**Comment by endorama** [comment]
@felixfontein I added the tests, I'm uncertain on which cases should I test (as testing `sops` behaviour itself seems redundant); I added a "missing file", "wrong file" and "success" tests, let me know if you were thinking of some other test cases.

File: lib/ansible/module_utils/sops.py
Comment (felixfontein) [inline lib/ansible/module_utils/sops.py:6]: Hmm, this is in fact a problem. Everything in `module_utils` can only use stuff from `module_utils`, but not from the rest of Ansible. That's somewhat unfortunate.

How about having a function `createSopsErrorMessage()` instead which is used by `SopsError` in the plugin to create the message? Then the heavy part (`sops_error_codes` and the logic to use it) is in `module_utils`, but `module_utils/sops.py` doesn't need to know about `AnsibleError`.

**Comment by felixfontein** [comment]
Please remove the binary file from the PR.

Also, you need a file `test/integration/targets/lookup_sops/aliases`, otherwise your test won't be run in CI. (See also the related `test/integration/targets/lookup_sops/aliases:0:0: missing alias 'shippable/posix/group[1-5]' or 'unsupported'` error above.)

**Comment by felixfontein** [comment]
I think the PR in ansible-collections/community.general#374 is now finished. Anyone interested in it, please take a look and check it out! Reviews and testing are always very welcome :)

**Comment by felixfontein** [comment]
The sops lookup plugin will be part of the next community.general release.

close_me

**Comment by endorama** [comment]
In case someone subscribed to this PR is interested, I started the process to create a new collection named `community.sops` where to move all sops related plugins and code. In case someone is interested in being a maintainer (I will) please get in touch in https://github.com/ansible-collections/overview/issues/85

**Pull request closed without merging**
